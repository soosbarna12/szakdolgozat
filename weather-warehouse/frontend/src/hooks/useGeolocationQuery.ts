import { useQuery } from "@tanstack/react-query";
import axios from "../utils/axiosConfig";
import { Pages } from "../types/page.type";


/*
export function useGeolocationQuery(location: string) {
    
  const { data, error, isLoading } = useQuery({
    queryKey: ["location", location],
    queryFn: async () => {
      const response = await axios.get(`/geo/location`, {
        params: { location }
      })

      if (!response?.data || response?.data?.length === 0) {
        return []; // return empty array if no data found
      }

      // search for unique locations
      const newResults = response.data.filter(
        (loc: any, index: number, self: any[]) =>
          index ===
          self.findIndex(
            (t) => t.name === loc.name && t.state === loc.state && t.country === loc.country
          )
      );
      return newResults; 
    },

    enabled: !!location,
  });

  return { data, error, isLoading };
}
*/

export function useGeolocationQuery(locationName: string, type: Pages) {
  const { data, error, isLoading } = useQuery({
    queryKey: ["locationName", locationName],
    queryFn: async () => {
      if (!locationName.trim()) {
        return []; // Return an empty array if no city name is provided
      }

      try {
        const response = await axios.get(`/geo/location`, {
          params: {
            location: locationName.trim(),
           },
        });

        if (!response?.data || response?.data?.length === 0) {
          return []; // Return an empty array if no data is found
        }

        // Filter unique city results
        const uniqueResults = response.data.filter(
          (loc: any, index: number, self: any[]) =>
            index ===
            self.findIndex(
              (t) => t.name === loc.name && t.state === loc.state && t.country === loc.country
            )
        );
        //console.log("Unique results:", uniqueResults); // Log the unique results for debugging

        return uniqueResults;
      } catch (error) {
        console.error("Error fetching geolocation data:", error);
        throw new Error("Failed to fetch geolocation data");
      }
    },
    enabled: (locationName.length !== 0 && type === Pages.Today),
    refetchOnWindowFocus: false, // Prevent refetching when the window regains focus
  });

  return { data, error, isLoading };
}