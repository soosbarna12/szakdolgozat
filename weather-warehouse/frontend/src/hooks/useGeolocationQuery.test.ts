import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import { Pages } from '../types/page.type';
import axios from '../utils/axiosConfig';
import { renderHookWithQueryClient } from '../utils/test/renderHookWithQueryClient';
import { useGeolocationQuery } from './useGeolocationQuery';

jest.mock('../utils/axiosConfig');

describe('useGeolocationQuery', () => {
  const mockLocationName = 'New York';
  const mockResponseData = [
    { name: 'New York', state: 'NY', country: 'US' },
    { name: 'New York', state: 'NY', country: 'US' }, // Duplicate to test filtering
    { name: 'Los Angeles', state: 'CA', country: 'US' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches useGeolocationQuery hook snapshot', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockResponseData });

    const { result } = renderHookWithQueryClient(() =>
      useGeolocationQuery(mockLocationName, Pages.Today)
    );
    
    await waitFor(() => {
      expect(result.current.data).not.toBeUndefined();
    });


    expect(result.current).toMatchSnapshot();
  });

  it('fetches geolocation data successfully and filters unique results', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockResponseData });

    const { result } = renderHookWithQueryClient(() =>
      useGeolocationQuery(mockLocationName, Pages.Today)
    );

    await waitFor(() => {
      expect(result.current.data).toEqual([
        { name: 'New York', state: 'NY', country: 'US' },
        { name: 'Los Angeles', state: 'CA', country: 'US' },
      ]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('returns an undefined array when no location name is provided', async () => {
    const { result } = renderHookWithQueryClient(() =>
      useGeolocationQuery('', Pages.Today)
    );

    await waitFor(() => {
      expect(result.current.data).toEqual(undefined);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('returns an empty array when no data is found', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: [] });

    const { result } = renderHookWithQueryClient(() =>
      useGeolocationQuery(mockLocationName, Pages.Today)
    );

    await waitFor(() => {
      expect(result.current.data).toEqual([]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('does not fetch data when type is not Pages.Today', async () => {
    const { result } = renderHookWithQueryClient(() =>
      useGeolocationQuery(mockLocationName, Pages.Historical)
    );

    await waitFor(() => {
      expect(result.current.data).toEqual([]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });

    expect(axios.get).not.toHaveBeenCalled();
  });
});