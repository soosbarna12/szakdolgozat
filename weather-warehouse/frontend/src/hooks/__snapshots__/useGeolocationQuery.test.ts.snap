// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGeolocationQuery matches useGeolocationQuery hook snapshot 1`] = `
{
  "data": [
    {
      "country": "US",
      "name": "New York",
      "state": "NY",
    },
    {
      "country": "US",
      "name": "Los Angeles",
      "state": "CA",
    },
  ],
  "error": null,
  "isLoading": false,
}
`;
